File: scripts/linter/src/rules/function-call-spacing.ts
import { LintRule, LintContext, LintViolation } from '../types';

// Macros/keywords that are exceptions - they don't follow the spacing rule
const EXCEPTION_KEYWORDS = new Set([
    // C++ keywords that look like function calls
    'if',
    'for',
    'while',
    'switch',
    'catch',
    'sizeof',
    'alignof',
    'decltype',
    'typeid',
    'noexcept',
    'void',       // Used in function pointer types: void(int)
    'return',
    'throw',
    'co_return',
    'co_yield',
    'co_await',
    // Preprocessor
    'defined',
    // Common macros
    'NOLINT',
    'Expects',
    'Ensures',
    'assert',
    'static_assert',
    // Test macros - handled by test-macro-spacing rule
    'TEST_CASE',
    'SUBCASE',
    'CHECK',
    'CHECK_FALSE',
    'REQUIRE',
    'REQUIRE_FALSE',
    'REQUIRE_THROWS',
    'REQUIRE_THROWS_AS',
    'REQUIRE_THROWS_WITH',
    'REQUIRE_THROWS_WITH_AS',
    'REQUIRE_NOTHROW',
    'CHECK_THROWS',
    'CHECK_THROWS_AS',
    'CHECK_THROWS_WITH',
    'CHECK_THROWS_WITH_AS',
    'CHECK_NOTHROW',
    'WARN',
    'WARN_FALSE',
    'INFO',
    'CAPTURE',
    'GENERATE',
    'SECTION',
    // Qt macros
    'Q_OBJECT',
    'Q_PROPERTY',
    'Q_SIGNAL',
    'Q_SLOT',
    'Q_EMIT',
    'Q_INVOKABLE',
    'Q_DECLARE_METATYPE',
    'Q_ENUM',
    'Q_FLAG',
    'Q_NAMESPACE',
    'emit',
    'signals',
    'slots',
]);

// Check if a line is a preprocessor directive or comment
function isPreprocessorOrComment(line: string): boolean {
    const trimmed = line.trim();
    return trimmed.startsWith('#') || trimmed.startsWith('//') || trimmed.startsWith('/*') || trimmed.startsWith('*');
}

// Check if we're inside a string literal at a given position
function isInsideString(line: string, position: number): boolean {
    let inString = false;
    let stringChar = '';

    for (let i = 0; i < position; i++) {
        const char = line[i];
        const prevChar = i > 0 ? line[i - 1] : '';

        if (!inString && (char === '"' || char === "'")) {
            inString = true;
            stringChar = char;
        } else if (inString && char === stringChar && prevChar !== '\\') {
            inString = false;
        }
    }

    return inString;
}

// Find the matching closing parenthesis, returns -1 if not found on this line
function findClosingParen(line: string, openParenIndex: number): number {
    let depth = 1;
    let i = openParenIndex + 1;

    while (i < line.length && depth > 0) {
        const char = line[i];

        // Handle string literals
        if (char === '"' || char === "'") {
            const quote = char;
            i++;
            while (i < line.length && line[i] !== quote) {
                if (line[i] === '\\') i++; // Skip escaped char
                i++;
            }
        } else if (char === '(') {
            depth++;
        } else if (char === ')') {
            depth--;
            if (depth === 0) {
                return i;
            }
        }
        i++;
    }

    return -1; // Not found on this line
}

// Check if this is a multi-line call where ( is at end of line
// e.g., func(\n    arg1,\n    arg2\n)
function isMultiLineCall(line: string, openParenIndex: number, closeParenIndex: number): boolean {
    if (closeParenIndex !== -1) {
        return false; // Closing paren is on same line, not multi-line
    }
    // Check if ( is followed only by whitespace/nothing
    const afterOpen = line.substring(openParenIndex + 1).trim();
    return afterOpen.length === 0;
}

// Check if parentheses contain arguments (non-empty, non-whitespace)
// For multi-line cases, we need to look at subsequent lines
function hasArguments(
    lines: string[],
    lineIndex: number,
    openParenIndex: number,
    closeParenIndex: number
): boolean {
    const line = lines[lineIndex];

    if (closeParenIndex !== -1) {
        // Single line case - easy to check
        const content = line.substring(openParenIndex + 1, closeParenIndex).trim();
        return content.length > 0;
    }

    // Multi-line case - closing paren not on this line
    // Check if there's content after ( on this line
    const afterOpen = line.substring(openParenIndex + 1).trim();
    if (afterOpen.length > 0) {
        return true; // There's content after ( on this line
    }

    // Check next line for content (handles multi-line function signatures)
    if (lineIndex + 1 < lines.length) {
        const nextLine = lines[lineIndex + 1].trim();
        // If next line has content and doesn't start with ), there are arguments
        if (nextLine.length > 0 && !nextLine.startsWith(')')) {
            return true;
        }
    }

    return false;
}

// Pattern to match identifier followed immediately by ( without space
const FUNC_NO_SPACE = /\b([a-zA-Z_][a-zA-Z0-9_]*)\(/g;

// Pattern to match identifier followed by space then (
const FUNC_WITH_SPACE = /\b([a-zA-Z_][a-zA-Z0-9_]*)\s+\(/g;

export const functionCallSpacingRule: LintRule = {
    name: 'function-call-spacing',
    description: 'Functions with arguments need space before (, zero-arg functions do not',

    check(context: LintContext): LintViolation[] {
        const violations: LintViolation[] = [];

        for (let i = 0; i < context.lines.length; i++) {
            const line = context.lines[i];
            const lineNumber = i + 1;

            if (isPreprocessorOrComment(line)) {
                continue;
            }

            // Check for functions WITHOUT space before ( that HAVE arguments (error)
            let match;
            FUNC_NO_SPACE.lastIndex = 0;

            while ((match = FUNC_NO_SPACE.exec(line)) !== null) {
                const funcName = match[1];
                const matchStart = match.index;
                const openParenIndex = matchStart + funcName.length;

                if (isInsideString(line, matchStart)) {
                    continue;
                }

                if (EXCEPTION_KEYWORDS.has(funcName)) {
                    continue;
                }

                const closeParenIndex = findClosingParen(line, openParenIndex);

                // Skip multi-line calls where ( is at end of line - these don't need space
                if (isMultiLineCall(line, openParenIndex, closeParenIndex)) {
                    continue;
                }

                if (hasArguments(context.lines, i, openParenIndex, closeParenIndex)) {
                    violations.push({
                        rule: 'function-call-spacing',
                        message: `Missing space before '(' in '${funcName}' (has arguments)`,
                        line: lineNumber,
                        column: openParenIndex + 1,
                        severity: 'error',
                    });
                }
            }

            // Check for functions WITH space before ( that have NO arguments (error)
            FUNC_WITH_SPACE.lastIndex = 0;

            while ((match = FUNC_WITH_SPACE.exec(line)) !== null) {
                const funcName = match[1];
                const matchStart = match.index;
                const fullMatch = match[0];
                const openParenIndex = matchStart + fullMatch.length - 1;

                if (isInsideString(line, matchStart)) {
                    continue;
                }

                if (EXCEPTION_KEYWORDS.has(funcName)) {
                    continue;
                }

                const closeParenIndex = findClosingParen(line, openParenIndex);

                if (!hasArguments(context.lines, i, openParenIndex, closeParenIndex)) {
                    violations.push({
                        rule: 'function-call-spacing',
                        message: `Unnecessary space before '(' in '${funcName}' (no arguments)`,
                        line: lineNumber,
                        column: openParenIndex + 1,
                        severity: 'error',
                    });
                }
            }
        }

        return violations;
    },
};

File: scripts/linter/rules/function_call_spacing.cpp
#include "../linter.hpp"

#include <cctype>
#include <unordered_set>

namespace wisdom_linter
{
namespace
{
    auto isPreprocessorOrComment( std::string_view line ) -> bool
    {
        size_t start = line.find_first_not_of( " \t" );
        if ( start == std::string_view::npos )
        {
            return false;
        }
        std::string_view trimmed = line.substr( start );
        return trimmed.starts_with( "#" ) || trimmed.starts_with( "//" ) ||
               trimmed.starts_with( "/*" ) || trimmed.starts_with( "*" );
    }

    class FunctionCallSpacingRule : public Rule
    {
    public:
        [[nodiscard]] auto name() const -> std::string_view override
        {
            return "function-call-spacing";
        }

        [[nodiscard]] auto description() const -> std::string_view override
        {
            return "Functions with arguments need space before (, zero-arg functions do not";
        }

        [[nodiscard]] auto check( const LintContext& context ) const
            -> std::vector<LintViolation> override
        {
            std::vector<LintViolation> violations;

            static const std::unordered_set<std::string> exception_keywords = {
                "if", "for", "while", "switch", "catch", "sizeof", "alignof", "decltype",
                "typeid", "noexcept", "void", "return", "throw", "co_return", "co_yield",
                "co_await", "defined", "NOLINT", "Expects", "Ensures", "assert", "static_assert",
                "TEST_CASE", "SUBCASE", "CHECK", "CHECK_FALSE", "REQUIRE", "REQUIRE_FALSE",
                "REQUIRE_THROWS", "REQUIRE_THROWS_AS", "REQUIRE_THROWS_WITH",
                "REQUIRE_THROWS_WITH_AS", "REQUIRE_NOTHROW", "CHECK_THROWS", "CHECK_THROWS_AS",
                "CHECK_THROWS_WITH", "CHECK_THROWS_WITH_AS", "CHECK_NOTHROW", "WARN", "WARN_FALSE",
                "INFO", "CAPTURE", "GENERATE", "SECTION",
                "CHECK_EQ", "CHECK_NE", "CHECK_GT", "CHECK_LT", "CHECK_GE", "CHECK_LE",
                "REQUIRE_EQ", "REQUIRE_NE", "REQUIRE_GT", "REQUIRE_LT", "REQUIRE_GE", "REQUIRE_LE",
                "WARN_EQ", "WARN_NE", "WARN_GT", "WARN_LT", "WARN_GE", "WARN_LE",
                "CHECK_UNARY", "CHECK_UNARY_FALSE",
                "REQUIRE_UNARY", "REQUIRE_UNARY_FALSE",
                "WARN_UNARY", "WARN_UNARY_FALSE",
                "WARN_THROWS", "WARN_THROWS_AS", "WARN_THROWS_WITH", "WARN_THROWS_WITH_AS",
                "WARN_NOTHROW",
                "CHECK_MESSAGE", "REQUIRE_MESSAGE", "WARN_MESSAGE",
                "MESSAGE", "FAIL", "FAIL_CHECK",
                "Q_OBJECT", "Q_PROPERTY", "Q_SIGNAL",
                "Q_SLOT", "Q_EMIT", "Q_INVOKABLE", "Q_DECLARE_METATYPE", "Q_ENUM", "Q_FLAG",
                "Q_NAMESPACE", "emit", "signals", "slots",
                "EM_ASM", "EM_ASM_PTR", "EM_ASM_INT", "EM_ASM_DOUBLE", "EM_ASM_ARGS",
                "lengthBytesUTF8", "stringToUTF8", "UTF8ToString", "_malloc", "_free",
            };

            auto isInsideString = []( const std::string& line, size_t position ) -> bool {
                bool in_string = false;
                char string_char = 0;

                for ( size_t i = 0; i < position; ++i )
                {
                    char c = line[i];
                    char prev = ( i > 0 ) ? line[i - 1] : 0;

                    if ( !in_string && ( c == '"' || c == '\'' ) )
                    {
                        in_string = true;
                        string_char = c;
                    }
                    else if ( in_string && c == string_char && prev != '\\' )
                    {
                        in_string = false;
                    }
                }
                return in_string;
            };

            auto findClosingParen = []( const std::string& line, size_t open_paren ) -> int {
                int depth = 1;
                size_t i = open_paren + 1;
                bool in_string = false;
                char string_char = 0;

                while ( i < line.size() && depth > 0 )
                {
                    char c = line[i];
                    char prev = ( i > 0 ) ? line[i - 1] : 0;

                    if ( !in_string && ( c == '"' || c == '\'' ) )
                    {
                        in_string = true;
                        string_char = c;
                    }
                    else if ( in_string && c == string_char && prev != '\\' )
                    {
                        in_string = false;
                    }
                    else if ( !in_string )
                    {
                        if ( c == '(' )
                        {
                            ++depth;
                        }
                        else if ( c == ')' )
                        {
                            --depth;
                            if ( depth == 0 )
                            {
                                return static_cast<int>( i );
                            }
                        }
                    }
                    ++i;
                }
                return -1;
            };

            auto hasArguments = [&]( size_t line_idx, size_t open_paren, int close_paren ) -> bool {
                const auto& line = context.lines[line_idx];

                if ( close_paren >= 0 )
                {
                    for ( size_t i = open_paren + 1; i < static_cast<size_t>( close_paren ); ++i )
                    {
                        if ( line[i] != ' ' && line[i] != '\t' )
                        {
                            return true;
                        }
                    }
                    return false;
                }

                for ( size_t i = open_paren + 1; i < line.size(); ++i )
                {
                    if ( line[i] != ' ' && line[i] != '\t' )
                    {
                        return true;
                    }
                }

                if ( line_idx + 1 < context.lines.size() )
                {
                    const auto& next_line = context.lines[line_idx + 1];
                    size_t start = next_line.find_first_not_of( " \t" );
                    if ( start != std::string::npos && next_line[start] != ')' )
                    {
                        return true;
                    }
                }

                return false;
            };

            auto isMultiLineCall = [&]( const std::string& line, size_t open_paren,
                                        int close_paren ) -> bool {
                if ( close_paren >= 0 )
                {
                    return false;
                }
                for ( size_t i = open_paren + 1; i < line.size(); ++i )
                {
                    if ( line[i] != ' ' && line[i] != '\t' )
                    {
                        return false;
                    }
                }
                return true;
            };

            for ( size_t i = 0; i < context.lines.size(); ++i )
            {
                const auto& line = context.lines[i];
                int line_number = static_cast<int>( i + 1 );

                if ( isPreprocessorOrComment( line ) )
                {
                    continue;
                }

                for ( size_t pos = 0; pos < line.size(); ++pos )
                {
                    if ( !std::isalpha( line[pos] ) && line[pos] != '_' )
                    {
                        continue;
                    }

                    if ( pos > 0 && ( std::isalnum( line[pos - 1] ) || line[pos - 1] == '_' ) )
                    {
                        continue;
                    }

                    size_t name_start = pos;
                    size_t name_end = pos;
                    while ( name_end < line.size() &&
                            ( std::isalnum( line[name_end] ) || line[name_end] == '_' ) )
                    {
                        ++name_end;
                    }

                    std::string func_name = line.substr( name_start, name_end - name_start );

                    if ( exception_keywords.count( func_name ) )
                    {
                        pos = name_end - 1;
                        continue;
                    }

                    if ( isInsideString( line, name_start ) )
                    {
                        pos = name_end - 1;
                        continue;
                    }

                    size_t after_name = name_end;
                    bool has_space = false;
                    while ( after_name < line.size() &&
                            ( line[after_name] == ' ' || line[after_name] == '\t' ) )
                    {
                        has_space = true;
                        ++after_name;
                    }

                    if ( after_name >= line.size() || line[after_name] != '(' )
                    {
                        pos = name_end - 1;
                        continue;
                    }

                    size_t open_paren = after_name;
                    int close_paren = findClosingParen( line, open_paren );

                    if ( isMultiLineCall( line, open_paren, close_paren ) )
                    {
                        pos = name_end - 1;
                        continue;
                    }

                    bool has_args = hasArguments( i, open_paren, close_paren );

                    if ( !has_space && has_args )
                    {
                        violations.push_back( LintViolation {
                            std::string { name() },
                            "Missing space before '(' in '" + func_name + "' (has arguments)",
                            line_number,
                            static_cast<int>( open_paren + 1 ),
                            Severity::Error,
                        } );
                    }
                    else if ( has_space && !has_args )
                    {
                        violations.push_back( LintViolation {
                            std::string { name() },
                            "Unnecessary space before '(' in '" + func_name + "' (no arguments)",
                            line_number,
                            static_cast<int>( open_paren + 1 ),
                            Severity::Error,
                        } );
                    }

                    pos = name_end - 1;
                }
            }

            return violations;
        }
    };
} // namespace

auto createFunctionCallSpacingRule() -> std::shared_ptr<Rule>
{
    return std::make_shared<FunctionCallSpacingRule>();
}

} // namespace wisdom_linter
